diff --git a/dist/esm/issuer.js b/dist/esm/issuer.js
index 1b18009bfeaa09f2222e9df760dde0e4eb1d4326..b9b8a2044cdb01cd873ecfa94ca049d640868c37 100644
--- a/dist/esm/issuer.js
+++ b/dist/esm/issuer.js
@@ -232,12 +232,18 @@ function issuer(input) {
     credentials: false
   }), async (c) => {
     const iss = issuer2(c);
+
+    const grant_types_supported = ["authorization_code", "refresh_token"];
+    if (Object.values(input.providers).some((provider) => provider.client !== undefined))
+      grant_types_supported.push("client_credentials");
+
     return c.json({
       issuer: iss,
       authorization_endpoint: `${iss}/authorize`,
       token_endpoint: `${iss}/token`,
       jwks_uri: `${iss}/.well-known/jwks.json`,
-      response_types_supported: ["code", "token"]
+      response_types_supported: ["code", "token"],
+      grant_types_supported,
     });
   });
   app.post("/token", cors({
@@ -292,6 +298,7 @@ function issuer(input) {
       }
       const tokens = await generateTokens(c, payload);
       return c.json({
+        token_type: "Bearer",
         access_token: tokens.access,
         expires_in: tokens.expiresIn,
         refresh_token: tokens.refresh
@@ -332,50 +339,67 @@ function issuer(input) {
         generateRefreshToken
       });
       return c.json({
+        token_type: "Bearer",
         access_token: tokens.access,
         refresh_token: tokens.refresh,
         expires_in: tokens.expiresIn
       });
     }
     if (grantType === "client_credentials") {
-      const provider = form.get("provider");
-      if (!provider)
-        return c.json({ error: "missing `provider` form value" }, 400);
-      const match = input.providers[provider.toString()];
-      if (!match)
-        return c.json({ error: "invalid `provider` query parameter" }, 400);
-      if (!match.client)
-        return c.json({ error: "this provider does not support client_credentials" }, 400);
+      let providerClient;
+      let providerName;
+      const requestedProviderName = form.get("provider");
+      if (requestedProviderName) {
+        providerName = requestedProviderName.toString();
+        const provider = input.providers[providerName];
+        if (!provider) return c.json({ error: "invalid 'provider' form value" }, 400);
+        if (!provider.client)
+          return c.json({ error: "requested provider does not support client_credentials" }, 400);
+        providerClient = provider.client;
+      } else {
+        const match = Object.entries(input.providers).find(
+          ([, provider]) => provider.client !== undefined
+        );
+        if (!match || !match[1].client)
+          return c.json({ error: "no provider supports client_credentials" }, 400);
+        providerName = match[0];
+        providerClient = match[1].client;
+      }
       const clientID = form.get("client_id");
       const clientSecret = form.get("client_secret");
       if (!clientID)
         return c.json({ error: "missing `client_id` form value" }, 400);
       if (!clientSecret)
         return c.json({ error: "missing `client_secret` form value" }, 400);
-      const response = await match.client({
+      const response = await providerClient({
         clientID: clientID.toString(),
         clientSecret: clientSecret.toString(),
         params: Object.fromEntries(form)
       });
       return input.success({
         async subject(type, properties, opts) {
-          const tokens = await generateTokens(c, {
-            type,
-            subject: opts?.subject || await resolveSubject(type, properties),
-            properties,
-            clientID: clientID.toString(),
-            ttl: {
-              access: opts?.ttl?.access ?? ttlAccess,
-              refresh: opts?.ttl?.refresh ?? ttlRefresh
-            }
-          });
+          const tokens = await generateTokens(
+            c,
+            {
+              type,
+              subject: opts?.subject || await resolveSubject(type, properties),
+              properties,
+              clientID: clientID.toString(),
+              ttl: {
+                access: opts?.ttl?.access ?? ttlAccess,
+                refresh: opts?.ttl?.refresh ?? ttlRefresh,
+              },
+            },
+            { generateRefreshToken: false},
+          );
           return c.json({
+            token_type: "Bearer",
             access_token: tokens.access,
-            refresh_token: tokens.refresh
+            expires_in: tokens.expiresIn
           });
         }
       }, {
-        provider: provider.toString(),
+        provider: providerName,
         ...response
       }, c.req.raw);
     }
diff --git a/src/issuer.ts b/src/issuer.ts
index 119a99a278d3fe05cab32c8e326a29fc848f0f6b..ce59db30cceaa2460ff5ec81859128c5b7316bfc 100644
--- a/src/issuer.ts
+++ b/src/issuer.ts
@@ -775,12 +775,18 @@ export function issuer<
     }),
     async (c) => {
       const iss = issuer(c)
+
+      const grant_types_supported = ["authorization_code", "refresh_token"]
+      if (Object.values(input.providers).some((provider) => provider.client !== undefined))
+        grant_types_supported.push("client_credentials")
+
       return c.json({
         issuer: iss,
         authorization_endpoint: `${iss}/authorize`,
         token_endpoint: `${iss}/token`,
         jwks_uri: `${iss}/.well-known/jwks.json`,
         response_types_supported: ["code", "token"],
+        grant_types_supported,
       })
     },
   )
@@ -879,6 +885,7 @@ export function issuer<
         }
         const tokens = await generateTokens(c, payload)
         return c.json({
+          token_type: "Bearer",
           access_token: tokens.access,
           expires_in: tokens.expiresIn,
           refresh_token: tokens.refresh,
@@ -947,6 +954,7 @@ export function issuer<
           generateRefreshToken,
         })
         return c.json({
+          token_type: "Bearer",
           access_token: tokens.access,
           refresh_token: tokens.refresh,
           expires_in: tokens.expiresIn,
@@ -954,24 +962,32 @@ export function issuer<
       }
 
       if (grantType === "client_credentials") {
-        const provider = form.get("provider")
-        if (!provider)
-          return c.json({ error: "missing `provider` form value" }, 400)
-        const match = input.providers[provider.toString()]
-        if (!match)
-          return c.json({ error: "invalid `provider` query parameter" }, 400)
-        if (!match.client)
-          return c.json(
-            { error: "this provider does not support client_credentials" },
-            400,
+        let providerClient: NonNullable<Provider<any>["client"]>
+        let providerName: string
+        const requestedProviderName = form.get("provider")
+        if (requestedProviderName) {
+          providerName = requestedProviderName.toString()
+          const provider = input.providers[providerName]
+          if (!provider) return c.json({ error: "invalid 'provider' form value" }, 400)
+          if (!provider.client)
+            return c.json({ error: "requested provider does not support client_credentials" }, 400)
+          providerClient = provider.client
+        } else {
+          const match = Object.entries(input.providers).find(
+            ([, provider]) => provider.client !== undefined
           )
+          if (!match || !match[1].client)
+            return c.json({ error: "no provider supports client_credentials" }, 400)
+          providerName = match[0]
+          providerClient = match[1].client
+        }
         const clientID = form.get("client_id")
         const clientSecret = form.get("client_secret")
         if (!clientID)
           return c.json({ error: "missing `client_id` form value" }, 400)
         if (!clientSecret)
           return c.json({ error: "missing `client_secret` form value" }, 400)
-        const response = await match.client({
+        const response = await providerClient({
           clientID: clientID.toString(),
           clientSecret: clientSecret.toString(),
           params: Object.fromEntries(form) as Record<string, string>,
@@ -979,25 +995,30 @@ export function issuer<
         return input.success(
           {
             async subject(type, properties, opts) {
-              const tokens = await generateTokens(c, {
-                type: type as string,
-                subject:
-                  opts?.subject || (await resolveSubject(type, properties)),
-                properties,
-                clientID: clientID.toString(),
-                ttl: {
-                  access: opts?.ttl?.access ?? ttlAccess,
-                  refresh: opts?.ttl?.refresh ?? ttlRefresh,
+              const tokens = await generateTokens(
+                c,
+                {
+                  type: type as string,
+                  subject:
+                    opts?.subject || (await resolveSubject(type, properties)),
+                  properties,
+                  clientID: clientID.toString(),
+                  ttl: {
+                    access: opts?.ttl?.access ?? ttlAccess,
+                    refresh: opts?.ttl?.refresh ?? ttlRefresh,
+                  },
                 },
-              })
+                { generateRefreshToken: false},
+              )
               return c.json({
+                token_type: "Bearer",
                 access_token: tokens.access,
-                refresh_token: tokens.refresh,
+                expires_in: tokens.expiresIn,
               })
             },
           },
           {
-            provider: provider.toString(),
+            provider: providerName,
             ...response,
           },
           c.req.raw,
